CH03-IOCoreSSL/CH03S01-Basic-NPN-and-ALPN.en.md

```diff
@@ -0,0 +1,199 @@
+# Basic Knowledge: NPN and APLN
+Before you read the implementation of SSL in ATS, Let's learn about NPN and APLN first. They are supposed to solve the same issue:
+  - Which protocol will be performed over plaintext after SSL connection is established?
+In the OSI model, SSL, widely used in many scenarios, is generally considered in the 6th layer. However,
+  - which protocol will be applied to the decrypted pipe, and how to apply it?
+  - It can only be decided by an agreement between the client and the server.
+  - For instance, on port 443, it's decided that we use SSL to encrypt traffic in HTTP/1.x.
+
+As SPDY is created, a requirement that using several possible application layer protocols on the same port has been put forward,
+
+  - We can't tell which protocol is applied unless data gets decrypted.
+  - We need to know which protocol before data gets decrypted.
+  - Therefore TLS-NPN extension is implemented.
+
+TLS-NPN extension implements that in SSL handshake:
+
+  - The server tells the client what protocols it supports.
+  - Then the client tells the server which protocol will be applied on this connection.
+
+TLS-NPN is short for Transport Layer Security - Next Protocol Negotiation (instead of NPN bipolar transistor, haha), it is implemented as an extension of TLS by Google to support SPDY as an application layer protocol.
+
+
+## TLS-NPN extension
+
+The detailed flow for a full SSL handshake with NPN is as follows. Please refer to [RFC 5246 Section 7.3](https://tools.ietf.org/html/rfc5246#section-7.3)
+
+```
+Client                                               Server
+
+ClientHello (NPN extension)   -------->
+                                               ServerHello
+                                                 (NPN extension &
+                                                  list of protocols)
+                                              Certificate*
+                                        ServerKeyExchange*
+                                       CertificateRequest*
+                            <--------      ServerHelloDone
+Certificate*
+ClientKeyExchange
+CertificateVerify*
+[ChangeCipherSpec]
+EncryptedExtensions
```

Comment:
I'm leave comments at Line 42 and Line 66 to describe the EncryptedExtensions.

The diagram is copied from the draft and modified by me to explain the different between NPN and ALPN. But you just copied the diagram from the draft and lost my modification.

I think that you need to understand them in detail in order to translate it into english.

Please just ignore the dup reference and translate into english depend on your understand but do not lost any things from diagram.